userspace/libsinsp/filterchecks.cpp

```diff
@@ -1784,6 +1841,86 @@ bool sinsp_filter_check_fd::compare_domain(sinsp_evt *evt)
 
 	return false;
 }
+
+bool sinsp_filter_check_fd::compare_fdtypes(sinsp_evt *evt)
+{
+	m_tinfo = evt->get_thread_info();
+	if(m_tinfo == NULL)
+	{
+		return false;
+	}
+
+	if (m_argid != -1) {
+		// We are of the form fd.types[num] so compare the specific
+		// file descriptor in num
+		m_fdinfo = m_tinfo->get_fd(m_argid);
+
+		return m_fdinfo ? flt_compare(m_cmpop,
+					      PT_CHARBUF,
+					      m_fdinfo->get_typestring()) : false;
+	}
+	else
+	{
+		// We are of the form fd.types so behave as a list of types for
+		// all open file descriptors
+		m_tinfo = evt->get_thread_info();
+
+		if(m_tinfo == NULL)
+		{
+			return false;
+		}
+
+		// All of the pointers come from the fd_typesting() function so
+		// we shouldn't have the situation of two distinct pointers to
+		// the same string literal
+		std::unordered_set<char*> fd_types;
+		bool result = false;
+
+		// Iterate over the list of open file descriptors and compare
+		// against m_val_storages_members, using a set to avoid
+		// comparing the same types more than once
+		auto fd_type_cmp = [this, &fd_types, &result](uint64_t, const sinsp_fdinfo_t& fdinfo)
+		{
+			char* type = fdinfo.get_typestring();
+
+			if (!fd_types.emplace(type).second)
+			{
+				for (const auto& m : m_val_storages_members)
+				{
+					if (::flt_compare(CO_EQ, PT_CHARBUF, type, m.first, strlen(type), m.second))
+					{
+						result = true;
+						return false;
+					}
+				}
+			}
+			return true;
+		};
+
+
+		// Would use sinsp_filter_check::flt_compare here but it requires
```

Comment:
You're right. Why not changing the signature of sinsp_filter_check::flt_compare into:
```
bool sinsp_filter_check::flt_compare(cmpop op, ppm_param_type type, void* operand1, uint32_t op1_len, uint32_t op2_len, bool is_list=false)
```

That way you could simply re-use the function here, and pass `m_info.m_fields[m_field_id].m_flags & EPF_IS_LIST` from the base-level `compare(...)` method.

Conversely, have you thought about the possibility of making this field a list instead of having it hybrid? The only limitation would be that even the `fd.types[<num>]` would be restricted to the `in`, `exists`, and `intersects` operators, but maybe it would be less ambiguous. Not strong here, just wanted to propose the idea.